### ISSUE TITLE
OpenVPN Client (OpenVPN GUI) only works with administrator rights

### ISSUE THREAD

**Issue by roswitina** [OpenVPN/openvpn#629]
I am currently using Windows 11 23H2 (Build 22631.4317) and have installed OpenVPN-2.6.12-I002-amd64. The OpenVPN GUI is loaded and the configurations are displayed. When I now start a configuration, I get the following error message in the log and the connection is cancelled.
```
2024-10-20 16:17:22 TCP/UDP: Preserving recently used remote address: [AF_INET]1.2.3.4:1194
2024-10-20 16:17:22 CreateFile failed on ovpn-dco device: \\?\ROOT#NET#0002#{cac88484-7515-4c03-82e6-71a87abac361}\ovpn-dco: access denied (errno=5)
2024-10-20 16:17:22 MANAGEMENT: Client disconnected
2024-10-20 16:17:22 All ovpn-dco adapters on this system are currently in use or disabled.
2024-10-20 16:17:22 Exiting due to fatal error
```
I remember that some time ago you had to start the OpenVPN client (GUI) with administrator rights for it to work, but I think that this was no longer necessary some time ago.

So I started the OpenVPN GUI with administrator rights and the connection setup worked again.

Am I wrong here and have I always had to start with administrator rights, or is there something wrong with my system? Where can I start?

**Comment by selvanair** [comment]
> Am I wrong here and have I always had to start with administrator rights, or is there something wrong with my system? Where can I start?

No, its not advisable nor required to run OpenVPN-GUI as administrator. Could you check whether OpenVPNServiceInteractive is running? It should be running for the GUI to work correctly. This service is installed and started by default, unless you manually disabled it. Please check.

**Comment by lstipakov** [comment]
We've have seen some [similar](https://github.com/OpenVPN/ovpn-dco-win/issues/38) reports. Unfortunately I don't have a clear steps to reproduce and a solution for this problem. Apparently there is something on machine which triggers this behavior. It has been reported that when running `msiexec /i OpenVPN-2.6.12-I002-amd64.msi` under elevated privileges, the issue doesn't reproduce. Also you might want to add `disable-dco` and see if it works while we're investigating it.

**Comment by roswitina** [comment]
> > Am I wrong here and have I always had to start with administrator rights, or is there something wrong with my system? Where can I start?
> 
> No, its not advisable nor required to run OpenVPN-GUI as administrator. Could you check whether OpenVPNServiceInteractive is running? It should be running for the GUI to work correctly. This service is installed and started by default, unless you manually disabled it. Please check.

both services are running (OpenVPNServiceInteractive, OpenVPNService)

**Comment by selvanair** [comment]
> both services are running (OpenVPNServiceInteractive, OpenVPNService)

Thanks for checking.
I was not aware of the issue @lstipakov mentioned above. Try reinstalling from an elevated command prompt.

**Comment by roswitina** [comment]
> We've have seen some [similar](https://github.com/OpenVPN/ovpn-dco-win/issues/38) reports. Unfortunately I don't have a clear steps to reproduce and a solution for this problem. Apparently there is something on machine which triggers this behavior. It has been reported that when running `msiexec /i OpenVPN-2.6.12-I002-amd64.msi` under elevated privileges, the issue doesn't reproduce. Also you might want to add `disable-dco` and see if it works while we're investigating it.

I have run `msiexec /i OpenVPN-2.6.12-I002-amd64.msi` in a console with administrator rights. Everything worked immediately after the installation. However, after logging out and logging in again, it no longer worked. In addition, I was only able to install the `OpenVPN-2.6.12-I001-amd64.msi` version. I was unable to download the `OpenVPN-2.6.12-I002-amd64.msi` version.

I also use the entry windows-driver wintun in my conf file. The following settings worked

**worked**
```
#OpenVPN Client conf
disable-dco
#windows-driver wintun
tls-client
client
...
```

**does not work**
```
#OpenVPN Client conf
#disable-dco
#windows-driver wintun
tls-client
client
...
```

**does not work**
```
#OpenVPN Client conf
disable-dco
windows-driver wintun
tls-client
client
...
```

**does not work**
```
#OpenVPN Client conf
#disable-dco
windows-driver wintun
tls-client
client
...
```

**Comment by lstipakov** [comment]
Obviously installing from elevated prompt cannot be considered as a fix and should not matter. Something is going on with access control which we do not fully understand.

**Comment by Frau-Mustermann** [comment]
Mir geht es genauso. Habe OpenVPN deinstalliert und neu installiert.  Trotzdem kann ich die Software nur mit Admi istratorrechten starten. Auf anderen Rechnern funktioniert es ohne Probleme. Wer kann hier helfen?

**Comment by lstipakov** [comment]
Would it be possible for you to provide a remote access to that machine to further debug this issue?

**Comment by irwir** [comment]
Installing alpha2 without GUI on the current Windows Canary in Virtualbox.
Disabled everything except service and DCO.
Interactive service had started while OpenVPNService failed.

Event Viewer shows in `System` time out events 7009 and 7000, and in `Application` :
> Event 1026
> Application: openvpnserv2.exe
> Framework Version: v4.0.30319
> Description: The process was terminated due to an unhandled exception.
> Exception Info: System.Security.SecurityException
>    at System.Diagnostics.EventLog.FindSourceRegistration(System.String, System.String, Boolean, Boolean)
>    at System.Diagnostics.EventLog.SourceExists(System.String, System.String, Boolean)
>    at System.Diagnostics.EventLog.SourceExists(System.String)
>    at OpenVpn.OpenVpnService..ctor()
>    at OpenVpn.Program.Main(System.String[])
> 
> Event 1000
> Faulting application name: openvpnserv2.exe, version: 2.0.0.0, time stamp: 0x67d2d744
> Faulting module name: KERNELBASE.dll, version: 10.0.27881.1000, time stamp: 0x392e7c44
> Exception code: 0xe0434352
> Fault offset: 0x00179554
> Faulting process id: 0xD14
> Faulting application start time: 0x1DBE4637FBEF3DF
> Faulting application path: C:\Program Files\OpenVPN\bin\openvpnserv2.exe
> Faulting module path: C:\WINDOWS\System32\KERNELBASE.dll
>

OpenVPNService would not start as neither Network nor Local service, only as System.

**Comment by lstipakov** [comment]
@irwir I cannot reproduce this. out of the box, apparently this code causes the issue (the first line):

```
if (!EventLog.SourceExists(this.ServiceName))
{
    EventLog.CreateEventSource(this.ServiceName, "Application");
}
``` 

Also this seem to be the separate issue from the one originally reported. I created a [new ticket](https://github.com/OpenVPN/openvpnserv2/issues/26) under openvpnser2 project.

**Comment by lstipakov** [comment]
@irwir Please check my comment under the new ticket.

**Comment by lstipakov** [comment]
Since there is no response from the original poster, I'll close the ticket. openvpnserv2 crash is handled in different issue.

**Comment by NIVA-IS-PSH** [comment]
I can reproduce what [roswitina](https://github.com/roswitina) is describing here: https://github.com/OpenVPN/openvpn/issues/629#issuecomment-2426980872.
Which basically is: dco and win-tun drivers only work if OpenVPN GUI is started with admin rights but TAP drivers work under normal user rights.  The two helper services are running and I have tried several reinstall etc. The version is: OpenVPN-2.6.14-I004-amd64

I'm able and willing to test stuff if needed. I can't provide remote access though.

**Comment by s31t0n** [comment]
I was able to reproduce this issue and implement a working workaround.

## Environment

- Windows 11 Pro 25H2 x64, OS Build 26200.8875
- OpenVPN Community 2.7.6
- OpenVPN GUI 11.65.0.0
- ovpn-dco-win 2.8.4
- OpenVPN GUI running as a standard, non-elevated user

OpenVPN reports the operating system internally as:

```text
Windows version: 10.0.26200, amd64
```

This is still Windows 11; Windows 11 retains the NT version number `10.0` for compatibility.

## Symptoms

The relevant error was:

```text
Using device interface: ...\ovpn-dco
CreateFile failed on ovpn-dco device: ... Access is denied. (errno=5)
Failed to open ovpn-dco adapter
```

The same OpenVPN profile worked correctly when OpenVPN GUI was started with **Run as administrator**.

Adding:

```text
disable-dco
```

also allowed OpenVPN GUI to connect as a standard user, but throughput was significantly lower.

This confirmed that the VPN profile, authentication, routing, server, and DCO driver itself were working. The failure occurred specifically when a non-elevated `openvpn.exe` process tried to open the ovpn-dco device interface.

## Troubleshooting already attempted

The following actions did not solve the problem:

- uninstalling OpenVPN;
- deleting all ovpn-dco adapter instances;
- removing every ovpn-dco driver package from the Windows Driver Store;
- rebooting;
- reinstalling OpenVPN;
- reinstalling ovpn-dco-win 2.8.4;
- recreating the DCO adapter;
- resetting the Windows network stack;
- verifying that `OpenVPNServiceInteractive` was running;
- verifying that the user was a member of the `OpenVPN Administrators` group.

After a complete uninstall, Driver Store cleanup, reboot, and clean reinstall, the newly created DCO device still returned:

```text
Access is denied. (errno=5)
```

The failing operation remained the `CreateFile()` call against the ovpn-dco device path.

## Root cause identified

The DCO driver was installed, loaded, and discoverable, but the effective per-device security descriptor did not allow the standard user's OpenVPN process to open the DCO device interface.

Reinstalling the driver did not correct that per-device ACL.

The workaround was to modify the security descriptor of the ovpn-dco Plug and Play device instance and grant access only to the current Windows user's SID.

## What the workaround script does

The PowerShell script must be run once as administrator.

It performs the following operations:

1. Stops running `openvpn.exe` and `openvpn-gui.exe` processes.

2. Checks whether a persistent DCO adapter exists by running:

   ```text
   tapctl.exe list --hwid ovpn-dco
   ```

3. If the requested adapter does not exist, creates one with:

   ```text
   tapctl.exe create --hwid ovpn-dco --name "OpenVPN DCO User"
   ```

4. Uses Windows SetupAPI to enumerate currently present devices in the Network device class:

   ```text
   {4d36e972-e325-11ce-bfc1-08002be10318}
   ```

5. Selects only devices whose hardware ID is exactly:

   ```text
   ovpn-dco
   ```

6. Reads the per-device security descriptor through:

   ```text
   SPDRP_SECURITY_SDS
   ```

   using `SetupDiGetDeviceRegistryPropertyW()`.

7. Saves the original device instance IDs and SDDL security descriptors to a JSON backup under:

   ```text
   C:\ProgramData\OpenVPN\DcoAclFix
   ```

8. Retrieves the SID of the current Windows user.

9. Adds the following access-control entry to the device DACL:

   ```text
   (A;;GRGWGX;;;<CURRENT_USER_SID>)
   ```

   This grants Generic Read, Generic Write, and Generic Execute access only to the current user.

10. Writes the updated security descriptor through:

    ```text
    SetupDiSetDeviceRegistryPropertyW(..., SPDRP_SECURITY_SDS, ...)
    ```

11. Restarts each affected DCO device with:

    ```text
    pnputil.exe /restart-device <INSTANCE_ID>
    ```

12. Optionally updates a specified `.ovpn` profile by:

    - removing `disable-dco`;
    - removing existing `dev-node` directives;
    - adding:

      ```text
      dev-node "OpenVPN DCO User"
      ```

The script does not grant device access to `Everyone`.

It does not modify the ACL of the `ovpn-dco` kernel service.

It modifies only the security descriptors of present Plug and Play devices whose hardware ID is exactly `ovpn-dco`.

## How to locate and verify the OpenVPN profile

Before running the script with `-ConfigPath`, first identify the exact `.ovpn` profile used by the failing connection.

The profile filename is normally visible in the OpenVPN log, for example:

```text
config = 'Client.ovpn'
```

It may also correspond to the connection name displayed by OpenVPN GUI.

Open **PowerShell as Administrator** and list the `.ovpn` profiles in the standard OpenVPN directories:

```powershell
$OpenVpnRoots = @(
    "$env:USERPROFILE\OpenVPN",
    "$env:ProgramFiles\OpenVPN"
) | Where-Object {
    Test-Path -LiteralPath $_
}

Get-ChildItem `
    -Path $OpenVpnRoots `
    -Filter "*.ovpn" `
    -File `
    -Recurse `
    -ErrorAction SilentlyContinue |
    Sort-Object FullName |
    Select-Object FullName, LastWriteTime
```

Find the profile whose filename matches the profile shown in the OpenVPN log or OpenVPN GUI.

Set its exact path in a variable:

```powershell
$ProfilePath = "C:\Path\To\Client.ovpn"
```

Verify that the file exists:

```powershell
Test-Path -LiteralPath $ProfilePath
```

The result must be:

```text
True
```

Review the relevant profile directives before making changes:

```powershell
Get-Content -LiteralPath $ProfilePath |
    Select-String -Pattern '^\s*(remote|dev|dev-node|disable-dco)\b'
```

Confirm that this is the correct VPN profile before continuing.

Create an additional manual backup:

```powershell
$ManualBackup = "$ProfilePath.manual-backup-$(Get-Date -Format 'yyyyMMdd-HHmmss')"

Copy-Item `
    -LiteralPath $ProfilePath `
    -Destination $ManualBackup

Write-Host "Manual profile backup: $ManualBackup"
```

The script also creates its own timestamped profile backup when `-ConfigPath` is used, but this manual backup provides an additional recovery copy.

## How to run the workaround

Save or download the script as:

```text
Fix-OpenVpnDcoAcl-English-Generic.ps1
```

For example, place it in the current user's Downloads directory.

Close OpenVPN GUI completely.

Open **PowerShell as Administrator** and define the script path:

```powershell
$ScriptPath = "$env:USERPROFILE\Downloads\Fix-OpenVpnDcoAcl-English-Generic.ps1"
```

Verify that the script exists:

```powershell
Test-Path -LiteralPath $ScriptPath
```

The result must be:

```text
True
```

Allow scripts only for the current PowerShell process and unblock the downloaded file:

```powershell
Set-ExecutionPolicy -Scope Process Bypass -Force
Unblock-File -LiteralPath $ScriptPath
```

Run the script and provide the previously verified OpenVPN profile:

```powershell
& $ScriptPath -ConfigPath $ProfilePath
```

Administrator privileges are required only while applying the device ACL change.

When `-ConfigPath` is used, the script:

- backs up the original device ACLs;
- creates the persistent DCO adapter if necessary;
- grants the current user's SID access to the DCO device;
- restarts the affected device;
- creates a timestamped backup of the `.ovpn` file;
- removes `disable-dco`;
- removes existing `dev-node` directives;
- adds:

  ```text
  dev-node "OpenVPN DCO User"
  ```

An automatically created profile backup will have a name similar to:

```text
Client.ovpn.bak-20260806-153000
```

After the script completes:

1. Close the elevated PowerShell window.
2. Start OpenVPN GUI normally.
3. Do not use **Run as administrator**.
4. Connect using the modified profile.

After applying this per-device ACL change, OpenVPN GUI was able to open the DCO device successfully as a standard user. DCO remained enabled and normal throughput was restored.

## Running the ACL fix without modifying a profile

The script can also be run without `-ConfigPath`:

```powershell
& $ScriptPath
```

In that mode, it changes the DCO device ACL but does not modify any `.ovpn` file.

The profile must then be edited manually:

1. Remove or comment out:

   ```text
   disable-dco
   ```

2. Remove any existing `dev-node` directive.

3. Add:

   ```text
   dev-node "OpenVPN DCO User"
   ```

Using `-ConfigPath` is safer when the exact profile has already been identified and verified because the script creates a timestamped backup before modifying it.

## ACL rollback

To restore the most recent DCO device ACL backup, close OpenVPN GUI, open **PowerShell as Administrator**, and run:

```powershell
$ScriptPath = "$env:USERPROFILE\Downloads\Fix-OpenVpnDcoAcl-English-Generic.ps1"

Set-ExecutionPolicy -Scope Process Bypass -Force
Unblock-File -LiteralPath $ScriptPath

& $ScriptPath -Rollback
```

The script restores the original per-device SDDL values from the newest backup under:

```text
C:\ProgramData\OpenVPN\DcoAclFix
```

It then restarts the affected DCO devices.

To see the available ACL backups:

```powershell
Get-ChildItem `
    -LiteralPath "C:\ProgramData\OpenVPN\DcoAclFix" `
    -Filter "DcoAclBackup-*.json" |
    Sort-Object LastWriteTime -Descending |
    Select-Object FullName, LastWriteTime
```

To restore a specific ACL backup:

```powershell
$AclBackup = "C:\ProgramData\OpenVPN\DcoAclFix\DcoAclBackup-YYYYMMDD-HHMMSS.json"

& $ScriptPath `
    -Rollback `
    -BackupFile $AclBackup
```

## Restoring the original OpenVPN profile

The ACL rollback does not automatically restore a modified `.ovpn` profile.

List the profile backups created by the script:

```powershell
$ProfileDirectory = Split-Path -Parent $ProfilePath
$ProfileFileName  = Split-Path -Leaf $ProfilePath

Get-ChildItem `
    -LiteralPath $ProfileDirectory `
    -Filter "$ProfileFileName.bak-*" |
    Sort-Object LastWriteTime -Descending |
    Select-Object FullName, LastWriteTime
```

Select the backup that should be restored:

```powershell
$ProfileBackup = "C:\Path\To\Client.ovpn.bak-YYYYMMDD-HHMMSS"
```

Restore it:

```powershell
Copy-Item `
    -LiteralPath $ProfileBackup `
    -Destination $ProfilePath `
    -Force
```

The separate manual backup created before running the script can also be restored in the same way.

After restoring the original profile, restart OpenVPN GUI normally.

## Conclusion

This appears to show that, on affected systems, reinstalling OpenVPN or ovpn-dco-win does not address the actual failure.

The driver is installed, loaded, and discoverable, but the effective per-device security descriptor prevents the intended non-elevated OpenVPN process from opening the DCO device interface.

This script is a workaround, not an upstream fix.

The driver installer or device creation path should ensure that newly created ovpn-dco device instances receive a security descriptor compatible with OpenVPN GUI's intended non-elevated execution model. Access could potentially be granted to the appropriate OpenVPN local group or through another suitably restricted device ACL.

[Fix-OpenVpnDcoAcl-English-Generic.zip](https://github.com/user-attachments/files/30792211/Fix-OpenVpnDcoAcl-English-Generic.zip)

**Comment by cron2** [comment]
I am hiding the multipage essay from @s31t0n as it very much looks like AI slop, and is not actually relevant here - the issue is "it does not work *with the iservice running*," where the question "can an unprivileged user open the DCO device or not?" is not relevant.

"Open the TAP device works" is also not really relevant, as elevated privileges are needed as soon as routes are to be installed - so the real question is "why does the iservice not work for the original reports", and we'd need openvpn logs (`verb 4`), gui logs, and system event logs to sort this out.
